### PR TITLE
Optional HSM slot index configuration

### PR DESCRIPTION
**Pull request by zpotoloom**
Implements https://github.com/vrk-kpa/xroad-joint-development/issues/177

### CODE REVIEW

File: doc/Manuals/ig-ss_x-road_v6_security_server_installation_guide.md [hunk 1]
```diff
@@ -6,7 +6,7 @@
 # Security Server Installation Guide
 **X-ROAD 6**
 
-Version: 2.9  
+Version: 3.0
```
Comment (VitaliStupin) [inline doc/Manuals/ig-ss_x-road_v6_security_server_installation_guide.md]: 2.10 would be a better version

File: doc/Manuals/ig-ss_x-road_v6_security_server_installation_guide.md [hunk 2]
```diff
@@ -245,6 +246,9 @@ To configure support for hardware security tokens (smartcard, USB token, Hardwar
 
         sudo service xroad-signer restart
 
+With some HSM drivers, it is not possible to configure specific slots for the server to use.
+To prevent scanning a lot of unnecessary slots specific slot indexes can be configured in `/etc/xroad/conf.d/signer.ini`
```
Comment (VitaliStupin) [inline doc/Manuals/ig-ss_x-road_v6_security_server_installation_guide.md]: signer.ini should not be modified by end user. /etc/xroad/conf.d/local.ini should be used instead

File: src/packages/default-configuration/signer.ini [hunk 3]
```diff
@@ -23,3 +23,6 @@ key-length=2048
 ; Certificate signing request signature digest algorithm,
 ; possible values: SHA-256, SHA-384, SHA-512
 csr-signature-digest-algorithm=SHA-256
+
+; Optional comma separated list of HSM slot indexes
+;hsm-slot-indexes=0,7,12
```
Comment (VitaliStupin) [inline src/packages/default-configuration/signer.ini]: There is no need to put this parameter to signer.ini. End users should add that to local.ini manually.

File: doc/Manuals/ig-ss_x-road_v6_security_server_installation_guide.md [hunk 4]
```diff
@@ -245,6 +246,14 @@ To configure support for hardware security tokens (smartcard, USB token, Hardwar
 
         sudo service xroad-signer restart
 
+With some HSM drivers, it is not possible to configure specific slots for the server to use.
+To prevent scanning a lot of unnecessary slots specific slot indexes for signer to use can be
+configured in `/etc/xroad/conf.d/local.ini`. For example:
+`[signer]
```
Comment (VitaliStupin) [inline doc/Manuals/ig-ss_x-road_v6_security_server_installation_guide.md]: Visual bug. All lines are collapsed into one line in github.

**Review by VitaliStupin — APPROVED** [review]

File: src/addons/hwtoken/src/main/java/ee/ria/xroad/signer/tokenmanager/module/HardwareModuleWorker.java [hunk 5]
```diff
@@ -109,19 +110,43 @@ protected void deinitializeModule() throws Exception {
...
-            if (previous == null) {
-                log.info("Module '{}' slot #{} has token with ID '{}': {}", module.getType(), slotIndex, token.getId(),
-                        token);
-            } else {
-                log.info("Module '{}' slot #{} has token with ID '{}' but token with that ID is already registered",
-                        module.getType(), slotIndex, token.getId());
+        // HSM slots defined in signer configuration
+        String[] hsmSlots = SystemProperties.getHSMSlotIndexes();
+        // Scan all available HSM slots
+        if (hsmSlots[0] == "") {
```
Comment (jhyoty) [inline src/addons/hwtoken/src/main/java/ee/ria/xroad/signer/tokenmanager/module/HardwareModuleWorker.java]: Strings should not be compared using reference equality. Preferably, getHSMSlotIndexes should return a zero-length array if the property is not set.

File: src/addons/hwtoken/src/main/java/ee/ria/xroad/signer/tokenmanager/module/HardwareModuleWorker.java [hunk 6]
```diff
@@ -109,19 +110,43 @@ protected void deinitializeModule() throws Exception {
...
+                    token.getId(), token);
+                } else {
+                    log.info("Module '{}' slot #{} has token with ID '{}' but token with that ID is already registered",
+                            module.getType(), slotIndex, token.getId());
+                }
+            }
+        // Only scan defined slot if HSM_SLOT_INDEXES variable configured
+        } else {
+            for (int i = 0; i < hsmSlots.length; i++) {
+                if (hsmSlots[i].matches("\\d+")) {
```
Comment (jhyoty) [inline src/addons/hwtoken/src/main/java/ee/ria/xroad/signer/tokenmanager/module/HardwareModuleWorker.java]: It probably should be an error if an index is not a positive integer, or is larger than the number of slots. Note that parseInt throws NumberFormatException if the value exceeds Integer.MAX_VALUE.

File: src/addons/hwtoken/src/main/java/ee/ria/xroad/signer/tokenmanager/module/HardwareModuleWorker.java [hunk 7]
```diff
@@ -109,19 +110,43 @@ protected void deinitializeModule() throws Exception {
...
+                    int hsmSlotIndex = Integer.parseInt(hsmSlots[i]);
+                    if (hsmSlotIndex < slots.length) {
+                        TokenType token = createToken(slots, hsmSlotIndex);
+                        TokenType previous = tokens.putIfAbsent(token.getId(), token);
+
+                        if (previous == null) {
+                            log.info("Module '{}' slot #{} has token with ID '{}': {}", module.getType(), hsmSlotIndex,
+                            token.getId(), token);
+                        } else {
+                            log.info("Module '{}' slot #{} has token with ID '{}' "
```
Comment (jhyoty) [inline src/addons/hwtoken/src/main/java/ee/ria/xroad/signer/tokenmanager/module/HardwareModuleWorker.java:142]: Maybe it should be an error if the list of indexes contains duplicates?

File: src/common-util/src/main/java/ee/ria/xroad/common/SystemProperties.java [hunk 8]
```diff
@@ -760,6 +763,13 @@ public static int getSignerKeyLength() {
     }
 
     /**
+     * @return optional HSM device slot indexes to use, "" by default
+     */
+    public static String[] getHSMSlotIndexes() {
+        return System.getProperty(HSM_SLOT_INDEXES, "").split(",");
```
Comment (jhyoty) [inline src/common-util/src/main/java/ee/ria/xroad/common/SystemProperties.java]: Consider returning an empty (zero-length) array or list if the property does not exist (or is an empty string).

**Review by jhyoty — CHANGES_REQUESTED** [review]

File: src/addons/hwtoken/src/main/java/ee/ria/xroad/signer/tokenmanager/module/HardwareModuleWorker.java (hunk 6 above)
Comment (zpotoloom) [reply src/addons/hwtoken/src/main/java/ee/ria/xroad/signer/tokenmanager/module/HardwareModuleWorker.java]: Currently most of the uses of parseInt from String are not checked for invalid values. Should be added globally.

Comment (jhyoty) [reply src/addons/hwtoken/src/main/java/ee/ria/xroad/signer/tokenmanager/module/HardwareModuleWorker.java]: Here, some invalid values are ignored and skipped, where others (too large for an int) will throw an exception and stop the processing. This is inconsistent, the check should either skip all invalid values or an exception should be thrown.

File: src/addons/hwtoken/src/main/java/ee/ria/xroad/signer/tokenmanager/module/HardwareModuleWorker.java [hunk 9]
```diff
@@ -109,19 +110,48 @@ protected void deinitializeModule() throws Exception {
...
+                    token.getId(), token);
+                } else {
+                    log.info("Module '{}' slot #{} has token with ID '{}' but token with that ID is already registered",
+                            module.getType(), slotIndex, token.getId());
+                }
+            }
+        // Only scan defined slots if HSM_SLOT_INDEXES variable configured
+        } else {
+            for (int i = 0; i < hsmSlots.length; i++) {
+                if (hsmSlots[i].matches("\\d+")) {
```
Comment (jhyoty) [inline src/addons/hwtoken/src/main/java/ee/ria/xroad/signer/tokenmanager/module/HardwareModuleWorker.java:132]: Some invalid values are ignored and skipped, where others (too large for an int) will throw an exception and stop the processing. This is inconsistent, the check should either skip all invalid values or an exception should be thrown.

**Comment by jhyoty** [comment]
@zpotoloom, this PR needs updating. It has a merge conflict and some requested changes.

**Pull request closed without merging**